File: debug.cpp
#include <iostream>
#include <iomanip>
#include <format>

#include "debug.hpp"

void disChunk(const Chunk &chunk, const std::string &name)
{
    std::cout << "=== " << name << " ===" << std::endl;

    for(size_t offset = 0; offset < chunk.size(); offset++)
    {
        offset = disInstr(chunk, offset);
    }
}

size_t disInstr(const Chunk &chunk, size_t offset)
{
    std::cout << std::setfill('0') << std::setw(4) << offset << " ";

    auto instruction = chunk[offset];

    switch(instruction)
    {
        case OP_RETURN:
            return simpleInstr("OP_RETURN", offset);
        default:
            throw std::invalid_argument(std::format("Unknown opcode {}", instruction));
    }
}

size_t simpleInstr(const std::string &name, size_t offset)
{
    std::cout << name << std::endl;
    return offset + 1;
}

File: malz-lang/src/debug.rs
use crate::chunk::Chunk;
use crate::chunk::OpCode;

pub fn dis_chunk(code_chunk: &Chunk, name: &str){
    println!("==== {} ====", &name);

    let mut offset = 0;

    while offset < code_chunk.code.len() {
        offset = dis_instr(&code_chunk, offset);
    }
}
pub fn dis_instr(chunk: &Chunk, offset: usize) -> usize{
    print!("{}", format!("{:0>4} ", offset.to_string()));
    print!("[ {} ] ", chunk.get_line(offset));
    
    match OpCode::try_from(chunk.code[offset]).unwrap() {
        OpCode::RETURN => simple_instr("RETURN", offset),
        OpCode::CONSTANT => constant_instr("CONSTANT", &chunk, offset),
        OpCode::CONSTANT_LONG => long_constant_instr("CONSTANT_LONG", &chunk, offset)
    }
}
pub fn simple_instr(name: &str, offset: usize) -> usize{
    println!("{}", &name);
    offset + 1
}

pub fn constant_instr(name: &str, chunk: &Chunk, offset: usize) -> usize{
    let constant = chunk.code[offset + 1];
    
    println!("{}", format!("{} {:0>4} ", &name, constant.to_string()));

    offset + 2
}

pub fn long_constant_instr(name: &str, chunk: &Chunk, offset: usize) -> usize{
    println!(
        "{}", 
        format!(
            "{} {:0>4} {:0>4} {:0>4}",
            &name,
            chunk.code[offset + 1].to_string(),
            chunk.code[offset + 2].to_string(),
            chunk.code[offset + 3].to_string()
        )
    );

    offset + 4
}
